exporter/collector/metrics.go

```diff
@@ -272,44 +360,252 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
+			// (the WAL library requires at least 1 entry).
+			// Doing so prevents double-exporting in the event of a collector restart.
+			emptyReq := &monitoringpb.CreateTimeSeriesRequest{}
+			bytes, bytesErr := proto.Marshal(emptyReq)
+			if bytesErr != nil {
+				return bytesErr
 			}
+
+			err = me.wal.Write(writeIndex+1, bytes)
 			if err != nil {
-				errs = append(errs, fmt.Errorf("failed to export time series to GCM: %+v", s))
+				return err
 			}
 		}
+		// move read index forward if non retryable error (or exported successfully)
+		err = me.wal.TruncateFront(readIndex + 1)
+		if err != nil {
+			return err
+		}
+		return nil
 	}
-	if len(errs) > 0 {
-		return multierr.Combine(errs...)
+	return err
+}
+
+// watchWAL watches the WAL directory for a write then returns to the
+// continuallyPopWAL() loop.
+func (me *MetricsExporter) watchWALFile(ctx context.Context) error {
+	me.goroutines.Add(1)
+	defer me.goroutines.Done()
+	walWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	err = walWatcher.Add(me.wal.path)
+
+	if err != nil {
+		return err
+	}
+	watchCh := make(chan error)
+	var wErr error
+	go func() {
+		defer func() {
+			watchCh <- wErr
+			close(watchCh)
+			walWatcher.Close()
+		}()
+
+		select {
+		case <-me.shutdownC:
+			return
+		case <-ctx.Done():
+			wErr = ctx.Err()
+			return
+
+		case event, ok := <-walWatcher.Events:
+			if !ok {
+				return
+			}
+			switch event.Op {
+			case fsnotify.Remove:
+				wErr = fmt.Errorf("WAL file deleted")
+			case fsnotify.Rename:
+				wErr = fmt.Errorf("WAL file renamed")
+			case fsnotify.Write:
+				wErr = nil
+			}
+
+		case watchErr, ok := <-walWatcher.Errors:
+			if ok {
+				wErr = watchErr
+			}
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) runWALReadAndExportLoop(ctx context.Context) {
+	defer me.goroutines.Done()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-me.shutdownC:
+			// do one last final read/export then return
+			// otherwise the runner goroutine could leave some hanging metrics unexported
+			for {
+				err := me.readWALAndExport(runCtx)
+				if err != nil {
+					if !errors.Is(err, wal.ErrOutOfRange) {
+						me.obs.log.Error(fmt.Sprintf("error flushing remaining WAL entries: %+v", err))
+					}
+					break
+				}
+			}
+			return
+		default:
+			err := me.readWALAndExport(ctx)
+			if err == nil {
+				continue
+			}
+			// ErrNotFound from wal.Read() means the index is either 0 or out of
+			// bounds (indicating we're probably at the end of the WAL). That error
+			// will trigger a file watch for new writes (below this). For other
+			// errors, fail.
+			if !errors.Is(err, wal.ErrNotFound) {
+				me.obs.log.Error(fmt.Sprintf("error reading WAL and exporting: %+v", err))
+			}
+
+			// Must have been ErrNotFound, start a file watch and block waiting for updates.
```

Comment:
The above `if` block doesn't break or return. Won't we get here even if `err` is another error?

Reply:
The idea was just to log errors that aren't `ErrNotFound`, since not found means we're at the end of the WAL and can be expected sometimes. added a comment explaining this